docs: semantic types (#56703)
metabase/metabase: docs/data-modeling/field-types.md

The main data types in Metabase:

| Data Type  | Example database types                      |
| ---------- | ------------------------------------------- |
| Numeric    | `INTEGER`, `FLOAT`                          |
| Temporal   | `DATE`, `TIMESTAMP`                         |
| Text       | `VARCHAR`, `TEXT`                           |
| Text-like  | MongoDB `BSONID`, Postgres `Enum`           |
| Boolean    | Boolean                                     |
| Collection | `JSON`, BigQuery `RECORD`, MongoDB `Object` |

Metabase currently does not support array types with any database. You'll only be able to use **Is empty** or **Is not empty** filters on columns containing arrays.

For some fields, you tell Metabase to [cast them to a different data type](#editing-data-and-semantic-types) (for example, text to date).

## Semantic types

Semantic types are extra flavor that you can add to a field to communicate meaning and enable [additional functionality](#additional-functionality-for-data-and-semantic-types). Available semantic types depend on the underlying data types.

> ### Semantic types for any field

Reviewer: This section could use some editing. E.g., line items don't need line breaks; category description is missing.

- Entity key

  Used to indicate that the field uniquely identifies each row. Could be a Product ID, serial number, etc.

- Foreign key

>   Used to refer to an Entity key of another table in order to connect data from different tables that are related. For example, in a Products table, you might have a Customer ID field that points to a Customers table, where Customer ID is the Entity key. If you want to use [linked filters on dashboards](../dashboards/linked-filters.md), you must set up foreign key relationships.

Reviewer: ```suggestion
  Used to refer to an entity key of another table in order to connect data from related tables. For example, in an Orders table, you might have a foreign key called `customer_id` that points to a Customers table's entity key, `id`. If you want to use [linked filters on dashboards](../dashboards/linked-filters.md), you must set up foreign key relationships.
```

- Category

### Semantic types for numeric fields

- Quantity
- Score
- Percentage
- Financial
  - Currency
  - Discount
  - Income
- Location
  - Latitude
  - Longitude

### Semantic types for temporal fields

- Creation date
- Creation time
- Creation timestamp
- Joined date
- Joined time
- Joined timestamp
- Birthday

### Semantic types for text fields

- Entity name
- Email

  - URL
  - Image URL
  - Avatar URL

- Category
- Name
- Title
- Product
- Source
- Location
  - City
  - State
  - Country
  - ZipCode

### Semantic types for collection fields

- Field containing JSON.

  See [Working with JSON](./json-unfolding.md).

## Editing data and semantic types

> Admins and people with [Manage table metadata permissions](../permissions/data.md#manage-table-metadata-permissions) can cast data types and edit semantic types in Admin Table Metadata settings.

Reviewer: ```suggestion
Admins, and people with [permission to manage table metadata](../permissions/data.md#manage-table-metadata-permissions), can cast data types and edit semantic types in the Admin setting's Table Metadata tab.
```

### Cast data types

> Data types can't be edited in Metabase directly, but certain data types can be [cast to different types types](./metadata-editing.md#casting-to-a-specific-data-type) to be read differently, like interpreting a numerical data type as a date format.

Reviewer: ```suggestion
Data types can't be edited in Metabase directly, but you can cast certain [data types to different types](./metadata-editing.md#casting-to-a-specific-data-type) so that, for example, Metabase will interpret a text data type as a date type.
```

> Changes made in Table Metadata apply across your entire Metabase. Metabase currently only supports casting to a datetime type in Metadata settings. However, if you you build a query in the query builder, in you can use type casting custom expressions like [`date()`](../questions/query-builder/expressions-list.md#date) or [`integer()`](../questions/query-builder/expressions-list.md#integer) to cast a string to a different type in your query.

Reviewer: Maybe a note about how models or viz settings can override types/effects?

### Semantic types don't change the data types

You can pick a semantic type compatible with the underlying data type in [table metadata settings](./metadata-editing.md#field-types)

> Semantic types add meaning shouldn't be used for type casting. For example, picking a semantic type of "Quantity" for a text field will not force Metabase to treat the field as numeric. Use Semantic types to add additional functionality for your fields, like formatting or visualizations.

Reviewer: ```suggestion
Semantic types only add meaning; they should NOT be used for type casting. For example, if you set a text field's semantic type to "Quantity", Metabase will still treat the field as a text field. Instead, apply semantic types to tell Metabase how to format or visualize the field (like telling Metabase that a numeric values represents a percentage).
```

## What data and semantic types enable

### Display format

Some semantic types change the way the data in the field is displayed.

Formatting setting from Table Metadata settings will be applied across your Metabase, but people can change them for individual charts.

| Semantic type          | Format                                                                                                                                                                                                                                                             |
| ---------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Percentage             | Displayed as percentage, for example 0.75 will be displayed as 75\%                                                                                                                                                                                                |

> | Currency               | On charts and in detail view, the values are prepended by the currency symbol, e.g. `$134.65`. In table view, the currency symbol is only displayed in the header by default, but you can change the metadata formatting settings to show the symbol on every row. |

Reviewer: ```suggestion
| Currency               | On charts and in detail view, the values are prepended by the currency symbol, e.g., `$134.65`. By default in the table view, the currency symbol is only displayed in the header, but you can change the metadata formatting settings to show the symbol for every row. |
```

> | Latitude/Longitude     | Displayed as coordinates, e.g. `0.00000000° N`                                                                                                                                                                                                                     |

Reviewer: ```suggestion
| Latitude/Longitude     | Displayed as coordinates, e.g., `0.00000000° N`                                                                                                                                                                                                                     |
```

| Email                  | Display as a `mailto` link                                                                                                                                                                                                                                         |
| URL                    | Can format as a clickable link                                                                                                                                                                                                                                     |
| Image URL              | Can display as an image. See table format settings LINK                                                                                                                                                                                                            |

> | Avatar URL             | Can display as avatar circle image. See table format settings LINK                                                                                                                                                                                                 |

Reviewer: Need links

| Field containing JSON  | In detail view, display as prettified JSON                                                                                                                                                                                                                         |
| Entity and Foreign key | Highlighted in table view                                                                                                                                                                                                                                          |

### Visualizations

> When you build a query in the query builder, Metabase will automatically choose the most suitable chart for you based on the data types and the semantic types of the field in the "Group by" (you can change the chart type later).

Reviewer: ```suggestion
When you create a question in the query builder, Metabase will automatically choose the most suitable chart for you based on the data types and the semantic types of the field in the "Group by" step (you can change the chart type later).
```
